src/ess/polarization/supermirror.py

```diff
@@ -10,10 +11,48 @@
 from .types import PlusMinus, PolarizingElement, TransmissionFunction
 
 
-class SupermirrorEfficiencyFunction(Generic[PolarizingElement]):
+class SupermirrorEfficiencyFunction(Generic[PolarizingElement], ABC):
+    """Base class for supermirror efficiency functions"""
+
+    @abstractmethod
     def __call__(self, *, wavelength: sc.Variable) -> sc.DataArray:
         """Return the efficiency of a supermirror for a given wavelength"""
-        raise NotImplementedError
+
+
+@dataclass
+class SecondOrderPolynomialEfficiency(SupermirrorEfficiencyFunction[PolarizingElement]):
+    """
+    Efficiency of a supermirror as a second-order polynomial
+
+    The efficiency is given by a * wavelength^2 + b * wavelength + c
+
+    Parameters
+    ----------
+    a:
+        Coefficient of the quadratic term, with unit of 1/angstrom^2
+    b:
+        Coefficient of the linear term, with unit of 1/angstrom
+    c:
+        Constant term, dimensionless
+    """
+
+    a: sc.Variable
+    b: sc.Variable
+    c: sc.Variable
+
+    def __post_init__(self):
+        if self.a.unit != sc.Unit('1/angstrom**2'):
```

Comment:
Maybe it's better like this, alternatively we could keep the parameters dimensionless and deal with the units in the `__call__` method. The advantage is that we / the users don't have to care about the units here.

Reply:
> Maybe it's better like this, not sure.

The core philosophy of Scipp is to avoid values without units wherever possible. It avoids really bad and hard to track-down bugs, as well as being self-documenting. Fewer bugs + documentation is what we chose over minor convenience.

Reply:
Oh but I didn't suggest not having units. I suggested to define the parameters as dimensionless.

Reply:
But sure, it's probably better like this.

Reply:
> Oh but I didn't suggest not having units. I suggested to define the parameters as dimensionless.

Don't the values of the parameters depend on their units? That is, if the user has in mind "b is 123/nm" and sets `b=123` we get nonsense?

Reply:
I thought the parameters here were mostly used in the fitting. And then it's quite nice to have dimensionless parameters (so we don't have to pass trivial initial guesses etc). But if they are to be set by a user it's probably better to have units.

Reply:
That is beside the point: The goal is *not* to have it "nice" or "convenient". We favor safe over convenient.

We can gain some convenience by attempting converting `a` and `b` to the inverse of the square and the inverse of the `wavelength`'s units. This would make it more flexible (can set params in `nm`, use `wavelength` in `Angstrom`), while keeping it safe.

Reply:
I agree that its better to have units! I had a look on the supermirror_test.py - how did you choose the unit acceptance now? As I see you test all combinations of units and running it doesnt yield any error, does that mean that the users can define any unit for a,b,c?

Reply:
Yes, they can use any units, as long as they have the correct powers, w.r.t. to `wavelength`.

Reply:
But you are also testing wrong powers, e.g. a, b, c all in 1/angstrom (first part of the test) - and that does not give an error?

Reply:
You mean code like this?
```python
    with pytest.raises(sc.UnitError, match=" to `dimensionless` is not valid"):
        eff = pol.SecondDegreePolynomialEfficiency(
            a=sc.scalar(1.0, unit='1/angstrom'),
            b=sc.scalar(1.0, unit='1/angstrom'),
            c=sc.scalar(1.0),
        )
        eff(wavelength=wav)
```
Note the `with pytest.raises(sc.UnitError, match=" to `dimensionless` is not valid"):` --- this checks that the code raises the expected exception.

Reply:
Ah ok, and the following code which in principle is correct from dimensions of a,b,c checks that it does not comply with some wavelength/time , i.e., for comparing a,b,c in their respective angstrom units with a "wavelength" in angstrom/s ?
`        eff(wavelength=wav)
    with pytest.raises(sc.UnitError, match=" to `dimensionless` is not valid"):
        eff = pol.SecondDegreePolynomialEfficiency(
            a=sc.scalar(1.0, unit='1/angstrom**2'),
            b=sc.scalar(1.0, unit='1/angstrom'),
            c=sc.scalar(1.0),
        )
        eff(wavelength=wav / sc.scalar(1.0, unit='s'))
`

Reply:
Correct!